internal/query/projection/milestones.go

```diff
@@ -0,0 +1,290 @@
+				crdb.Not(crdb.NewTextArrayContainsCond(MilestoneColumnIgnoreClientIDs, e.ApplicationID)),
+				crdb.NewIsNullCond(MilestoneColumnReachedDate),
+			},
+		))
+	}
+	return crdb.NewMultiStatement(e, statements...), nil
+}
+
+func (p *milestoneProjection) reduceInstanceRemoved(event eventstore.Event) (*handler.Statement, error) {
+	if _, err := assertEvent[*instance.InstanceRemovedEvent](event); err != nil {
+		return nil, err
+	}
+	return p.reduceReachedFunc(milestone.InstanceDeleted)(event)
+}
+
+func (p *milestoneProjection) reduceMilestonePushed(event eventstore.Event) (*handler.Statement, error) {
+	e, err := assertEvent[*milestone.PushedEvent](event)
+	if err != nil {
+		return nil, err
+	}
+	if e.MilestoneType != milestone.InstanceDeleted {
+		return crdb.NewUpdateStatement(
+			event,
+			[]handler.Column{
+				handler.NewCol(MilestoneColumnPushedDate, event.CreationDate()),
+			},
+			[]handler.Condition{
+				handler.NewCond(MilestoneColumnInstanceID, event.Aggregate().InstanceID),
+				handler.NewCond(MilestoneColumnType, e.MilestoneType),
+			},
+		), nil
+	}
+	return crdb.NewDeleteStatement(
+		event,
+		[]handler.Condition{
+			handler.NewCond(MilestoneColumnInstanceID, event.Aggregate().InstanceID),
+		},
+	), nil
+}
+
+func (p *milestoneProjection) reduceReachedIfUserEventFunc(msType milestone.Type) func(event eventstore.Event) (*handler.Statement, error) {
+	return func(event eventstore.Event) (*handler.Statement, error) {
+		if p.isSystemEvent(event) {
+			return crdb.NewNoOpStatement(event), nil
+		}
+		return p.reduceReachedFunc(msType)(event)
+	}
+}
+
+func (p *milestoneProjection) reduceReachedFunc(msType milestone.Type) func(event eventstore.Event) (*handler.Statement, error) {
+	return func(event eventstore.Event) (*handler.Statement, error) {
+		return crdb.NewUpdateStatement(event, []handler.Column{
+			handler.NewCol(MilestoneColumnReachedDate, event.CreationDate()),
+		},
+			[]handler.Condition{
+				handler.NewCond(MilestoneColumnInstanceID, event.Aggregate().InstanceID),
+				handler.NewCond(MilestoneColumnType, msType),
+				crdb.NewIsNullCond(MilestoneColumnReachedDate),
+			}), nil
+	}
+}
+
+func (p *milestoneProjection) reduceAppConfigAdded(event eventstore.Event, clientID string) (*handler.Statement, error) {
+	if !p.isSystemEvent(event) {
+		return crdb.NewNoOpStatement(event), nil
+	}
+	return crdb.NewUpdateStatement(
+		event,
+		[]handler.Column{
+			crdb.NewArrayAppendCol(MilestoneColumnIgnoreClientIDs, clientID),
+		},
+		[]handler.Condition{
+			handler.NewCond(MilestoneColumnInstanceID, event.Aggregate().InstanceID),
+			handler.NewCond(MilestoneColumnType, milestone.AuthenticationSucceededOnApplication),
+			crdb.NewIsNullCond(MilestoneColumnReachedDate),
+		},
+	), nil
+}
+
+func (p *milestoneProjection) isSystemEvent(event eventstore.Event) bool {
+	return event.EditorUser() == "" || event.EditorService() == "" || event.EditorService() == "SYSTEM"
```

Comment:
i believe this check won't work as expected.... had a look at my instance events (without Admin, Mgmt and Auth API): 
```
     editor_user     | editor_service
---------------------+-----------------
                     | System-API
  OIDC               |
  SAML               |
  <userID>           |
  system             | system
  oidc               |
  LOGIN              |
  SYSTEM             |
  NOTIFICATION       |
                     |
(10 rows)
```

Reply:
I changed it to the more safe variant

```go
func (p *milestoneProjection) isSystemEvent(event eventstore.Event) bool {
	if _, err := strconv.Atoi(event.EditorUser()); err == nil {
		return false
	}
	lowerEditorService := strings.ToLower(event.EditorService())
	return lowerEditorService == "" ||
		lowerEditorService == "system" ||
		lowerEditorService == "system-api"
}
```